javascript/app-discover/src/components/lists/EdgeList.utils.ts

```diff
@@ -0,0 +1,145 @@
+/*!
+ * Copyright (c) Microsoft. All rights reserved.
+ * Licensed under the MIT license. See LICENSE file in the project.
+ */
+
+import type { CausalDiscoveryConstraints } from '../../domain/CausalDiscovery/CausalDiscoveryConstraints.js'
+import type { Relationship } from '../../domain/Relationship.js'
+import {
+	hasInvertedSourceAndTarget,
+	hasSameReason,
+	hasSameSourceAndTarget,
+	invertRelationship,
+	ManualRelationshipReason,
+} from '../../domain/Relationship.js'
+
+export function removeEdge(
+	constraints: CausalDiscoveryConstraints,
+	onUpdateConstraints: (newConstraints: CausalDiscoveryConstraints) => void,
+	relationship: Relationship,
+) {
+	const newConstraints = {
+		...constraints,
+		manualRelationships: [
+			...constraints.manualRelationships.filter(
+				r =>
+					!hasSameSourceAndTarget(r, relationship) ||
+					(hasSameReason(ManualRelationshipReason.Flipped, relationship) &&
+						!hasInvertedSourceAndTarget(r, relationship)),
+			),
+			{
+				...relationship,
+				reason: ManualRelationshipReason.Removed,
+			},
+		],
+	}
+	onUpdateConstraints(newConstraints)
+}
+export function pinEdge(
+	constraints: CausalDiscoveryConstraints,
+	onUpdateConstraints: (newConstraints: CausalDiscoveryConstraints) => void,
+	relationship: Relationship,
+) {
+	const isPinned = constraints.manualRelationships.find(
+		x => x.key === relationship.key,
+	)
+	const filtered = constraints.manualRelationships.filter(
+		x => x.key !== relationship.key,
+	)
+	const newConstraints = {
+		...constraints,
+		manualRelationships: isPinned
+			? filtered
```

Comment:
I don't quite get the logic here - if it's not pinned we use a reason called "pinned"? Should this be !isPinned?

Reply:
If it's not pinned we use a reason to pin it, because if it's pinned we just return the filtered list that removes the pinned item (the 'undo')

Reply:
riiiiiight I see it. I changed it to shouldUndo, this variable identifies if the constraint already exists on the list or not, because if it exists, we should use the filtered list that removes it